### PR TITLE
fix: don't send Disconnect until peer has been removed from peer_map

### PR DESCRIPTION
**Pull request by mattyg**
This is 2 changes to reduce test flakes of `sending_second_message_after_remote_disconnect_succeeds_after_disconnect_event`.

This PR seems to eliminate the flakes completely. On `main` it consistently failed with 10 test runs. In this PR it does not arise within 500 test runs with either backend. The test run that reproduced it was `sending_second_message_after_remote_disconnect_succeeds_after_disconnect_event`.

The changes are explained in inline comments.

## Summary by CodeRabbit

* **Bug Fixes**
  * Peer disconnects now use a 10-second timeout and log failures/timeouts to improve reliability and diagnostics.
  * Cleanup of peer connections is now driven by an explicit guard to avoid premature or immediate disconnects, reducing race conditions during shutdown and tests.

### CODE REVIEW

File: crates/tx5-connection/src/hub.rs [hunk 1]
```diff
@@ -239,14 +239,15 @@ impl Hub {
                         }
                     }
                     HubCmd::Disconnect(pub_key) => {
-                        // disconnect from a remote peer
+                        // Remove remote peer from hub map
```
Comment (mattyg) [inline crates/tx5-connection/src/hub.rs]: First remove the peer from the hub map, then call close_peer.

If `DynBackEp::connect` is called before this `close_peer` future resolves, then that connection attempt would fail with an error "already connected".

Since we are ignoring the result from `close_peer` here anyway, this does not change the logic.

File: crates/tx5/src/peer.rs [hunk 2]
```diff
@@ -339,23 +360,5 @@ async fn task(
             .await;
     }
 
-    // the above loop ended, notify of disconnect
-    match tokio::time::timeout(
```
Comment (mattyg) [inline crates/tx5/src/peer.rs]: By sending the disconnect here, it is sent before the peer is removed from the peer map.

Thus we may remove close the sbd connection for the peer and remove it from the hub map, but still have a Peer available that can be interacted with. Those interactions will fail.

Instead this change is to only send the disconnect when the Peer is dropped.

File: crates/tx5-connection/src/conn.rs [hunk 3]
```diff
@@ -238,12 +238,28 @@ impl TaskCore {
...
 }
+/// Drop guard that closes the ready semaphore. This ensures callers will not be waiting
+/// for ready indefinitely after the task has ended.
+struct DropReady {
+    ready: Arc<tokio::sync::Semaphore>,
+}
+
+impl Drop for DropReady {
+    fn drop(&mut self) {
+        self.ready.close();
```
Comment (ThetaSinner) [inline crates/tx5-connection/src/conn.rs]: I don't think it's acceptable to close this if a ready state is not reached. Closing the ready semaphore allows the code to proceed as though the connection can be used and we then rely on something downstream of the ready check failing and allowing everything to shut down. If nothing else, that is hard to reason about when reading the code.

The failure to ready is already handled in https://github.com/holochain/tx5/blob/main/crates/tx5/src/backend/be_tx5_connection.rs#L58 which will catch the case that something fails. It's not ideal and not as sophisticated as the solution higher up the stack. Changing that though, I think should look like communicating success or failure with the semaphore close, and not just closing it and letting the code proceed as though the connection is ready.

File: crates/tx5/src/peer.rs [hunk 4]
```diff
@@ -339,23 +360,10 @@ async fn task(
...
-        }
-    }
-
-    // all other cleanup is handled by the drop guard
+    // Drop the drop guard to run cleanup tasks.
+    //
+    // If the Peer is dropped early, then the DropPeer guard will be dropped with it.
+    // Otherwise if the above loop ends before the Peer is dropped,
+    // we drop the DropPeer guard manually to run cleanup tasks.
+    drop(drop_guard);
```
Comment (ThetaSinner) [inline crates/tx5/src/peer.rs:373]: Should not be necessary to drop explicitly here, but the expanded comment is useful

**Review by ThetaSinner — CHANGES_REQUESTED** [review]
I think the change to peer.rs makes sense, I don't think the change to conn.rs does. Does one work without the other? Dealing with the ready success/failure states is a bigger change

File: crates/tx5/src/peer.rs (hunk 4 above)
Comment (mattyg) [reply crates/tx5/src/peer.rs:373]: I agree, but without this a test does consistently fail within ~250 runs. With this it doesn't. I can't explain why, but could dig deeper to find out. Or we could keep in and I could note that in a comment. Or we could remove it and not worry about the edge case failure.

Comment (neonphog) [reply crates/tx5/src/peer.rs:373]: I also can't imagine why an explicit drop helps, but I totally believe you. Yes, I'd say add a comment to that effect and leave this in.

File: crates/tx5-connection/src/conn.rs (hunk 3 above)
Comment (neonphog) [reply crates/tx5-connection/src/conn.rs]: Indeed, making tasks wait a timeout when we know already it is not going to work isn't ideal...

It may be worth a quick spike to make the ready function (line 153) return a result to see how complicated an update that is. If it's not bad, the ready semaphore could be changed to a tokio oneshot that returns a `Result<()>`.

To start, the drop case could return a disconnect error, and later on we could send back more detailed errors for the connection failure branches.

**Review by neonphog — CHANGES_REQUESTED** [review]
I think it's worth spending a little time to attempt making the ready return more semantic results. If that turns out to be a rabbit hole, I'm pretty comfortable with these changes, maybe just adding a comment that closing the ready without somehow better indicating it is an error case is not ideal.

**Comment by mattyg** [comment]
For the sake of not putting too much time into tx5 edge case issues, I'm going to keep in the `peer.rs` change with an explanation comment and remove the `conn.rs` change rather than do additional refactoring.

This significantly reduces test flakes such that it shouldn't disrupt our productivity, so good enough to move forward.

**Review by ThetaSinner — APPROVED** [review]
